astropy/coordinates/stokes_coord.py

```diff
@@ -0,0 +1,160 @@
+from collections import namedtuple
+from contextlib import contextmanager
+from copy import copy
+from typing import Dict
+
+import numpy as np
+
+from astropy.utils import unbroadcast
+from astropy.utils.shapes import ShapedLikeNDArray
+
+__all__ = ["StokesCoord", "custom_stokes_symbol_mapping", "StokesSymbol"]
+
+
+StokesSymbol = namedtuple("StokesSymbol", ["symbol", "description"], defaults=[""])
+
+# This is table 29 in the FITS 4.0 paper
+FITS_STOKES_VALUE_SYMBOL_MAP = {
+    1: StokesSymbol("I", "Standard Stokes unpolarized"),
+    2: StokesSymbol("Q", "Standard Stokes linear"),
+    3: StokesSymbol("U", "Standard Stokes linear"),
+    4: StokesSymbol("V", "Standard Stokes circular"),
+    -1: StokesSymbol("RR", "Right-right circular"),
+    -2: StokesSymbol("LL", "Left-left circular"),
+    -3: StokesSymbol("RL", "Right-left cross-circular"),
+    -4: StokesSymbol("LR", "Left-right cross-circular"),
+    -5: StokesSymbol("XX", "X parallel linear"),
+    -6: StokesSymbol("YY", "Y parallel linear"),
+    -7: StokesSymbol("XY", "XY cross linear"),
+    -8: StokesSymbol("YX", "YX cross linear"),
+}
+
+STOKES_VALUE_SYMBOL_MAP = copy(FITS_STOKES_VALUE_SYMBOL_MAP)
+
+
+@contextmanager
+def custom_stokes_symbol_mapping(
+    mapping: Dict[int, StokesSymbol], replace: bool = False
+):
+    """
+    Add a custom set of mappings from values to Stokes symbols.
+
+    Parameters
+    ----------
+    mappings
+        A list of dictionaries with custom mappings between values (integers)
+        and `.StokesSymbol` classes.
+    replace
+        Replace all mappings with this one.
+    """
+    global STOKES_VALUE_SYMBOL_MAP  # noqa: PLW0603
+
+    original_mapping = copy(STOKES_VALUE_SYMBOL_MAP)
+    if not replace:
+        STOKES_VALUE_SYMBOL_MAP = {**original_mapping, **mapping}  # noqa: PLW0603
+    else:
+        STOKES_VALUE_SYMBOL_MAP = mapping  # noqa: PLW0603
+
+    yield
+
+    STOKES_VALUE_SYMBOL_MAP = original_mapping
+
+
+class StokesCoord(ShapedLikeNDArray):
+    """
+    A representation of stokes coordinates with helpers for converting to profile names.
+
+    Parameters
+    ----------
+    stokes_values : array-like
+        The numeric values representing stokes coordinates.
+    """
+
+    def __init__(self, stokes, copy=False):
+        stokes = np.asanyarray(stokes)
+        if stokes.dtype.kind == "U":
+            self._data = self._from_symbols(stokes)
+        else:
+            self._data = stokes
+
+    @property
+    def shape(self):
+        return self._data.shape
+
+    @property
+    def value(self):
+        return self._data
+
+    def __array__(self):
+        return self._data
```

Comment:
I added this to make `u.Quantity(StokesCoord...))` work, but now `np.equal` is broken, presumably because it's calling this first? @mhvk any ideas?

Reply:
I've also had a great deal of trouble with `__array__`. Numpy seems to inconsistently use that instead of `__array_(u)fun(tion)__`.

Reply:
What breaks if you do not define this? My overall sense is that this is array-like, but not so array-y that most numpy functions/ufuncs make sense (e.g., addition and multiplication make no sense).

So, I think `==` should of course work, but it is not obvious one cares about `np.equal`.

Reply:
I think that we still want to be able to convert this back to a Quantity / array, mainly to help with backwards compatibility. I.e. we still want `u.Quantity(StokesCoord)` to work for people to "get back" the old behaviour.

Reply:
@mhvk Do you have any more thoughts on this. The current state is that one thing is broken with it and a different one is broken without it lol